### DISCUSSION TITLE
Windows AD Authentication ?

### DISCUSSION THREAD

**Discussion by boutaga** [burningalchemist/sql_exporter#461] [Q&A] [answered]
Hi, I would like to know if Active Directory authentication is some what supported for SQL Server connection ?
What would be the best way to secure the password rather than having it in plain text ? 
Would it be possible to use the run as Windows service option and make the SQL Exporter use the credential of the service account ?

Thanks for the good work !

**Comment by boutaga** [comment, marked as answer]
I actually found how to this in my environment. 

1. Create the service account in AD,
2. Create the Service on the target host (as demonstrated in the documentation)
3. Configure the service account credential to start the Service.
4. Provide necessary GRANT to the service account in SQL Server
5. modify the data_source_name line as such : 

`data_source_name: 'sqlserver://localhost:1433'`

I was trying to use Integrated Security=SSPI in the config but didn't work and when I tried without specifying anything it worked :)

**Comment by burningalchemist** [comment]
Hi @boutaga, my experience with modern Windows and AD is limited, and I don't have any Windows-based server environment at hand to recommend anything. But I'd follow this way:
- https://github.com/microsoft/go-mssqldb?tab=readme-ov-file#connection-parameters-and-dsn
The underlying driver supports many authentication options (e.g. one of our users successfully logged in using Azure AD), which are configured via parameters. So I assume, if you provide these parameters correctly (including `user id`, `authenticator`, maybe something else), this should work;
- there's `usql` CLI application (https://github.com/xo/usql) which you can use to connect to the database and debug connection erros. If it works with `usql`, it will work with `sql_exporter` too (the connection mechanism is the same).

As for Windows service option, in the readme you can find a quick and dirty way to run sql_exporter as a service. The better option would be to use NSSM (https://nssm.cc), which proved to work [here](https://github.com/burningalchemist/sql_exporter/issues/367#issuecomment-1795065285). Sadly, the solution was never shared. 🤷‍♂️

UPD: I see you found the solution. Thanks for sharing it! 👍 I'll use this information when I begin creating a documentation site. Cheers! :)
